rfcs/2023-01-06-layout.md

```diff
@@ -0,0 +1,461 @@
+| Position            | `position`, `zIndex`, `top`, etc.                                              | [List of props](https://github.com/styled-system/styled-system/blob/master/docs/table.md#position) referenced from styled-system.                                   |
+| Grid **(web only)** | `gridColumn*`, `gridRow*`, `gridArea`, `justifySelf`, `alignSelf`, `placeSelf` | Only restricted set of properties applicable on [grid children](https://css-tricks.com/snippets/css/complete-guide-grid/#aa-properties-for-the-childrengrid-items). |
+| Layout              | None                                                                           |                                                                                                                                                                     |
+| Gap                 | None                                                                           |                                                                                                                                                                     |
+
+#### `useSx` hook
+
+A `useSx` hook, inspired from [Dripsy](https://www.dripsy.xyz/apis/useSx#recommendations) can be helpful for targeting some minor usecases where users may want to style custom components or components which aren't directly accessible to be styled (for eg. `contentContainerStyle` on `FlatList`). I don't suspect a lot of usecases of this prop currently so this can be taken up later as en enhancement. Sample usage:
+
+```tsx
+import { useSx } from '@razorpay/blade/utils';
+
+// ...
+
+const sx = useSx();
+
+return (
+  <>
+    <CustomHeader style={sx({ marginBottom: 'spacing.2' })} />
+    <FlatList
+      contentContainerStyle={sx({ marginX: 'spacing.1' })}
+      data={data}
+      renderItem={renderItem}
+    />
+  </>
+);
+```
+
+`useSx` should return a function (named `sx` above) that can accept the same keys supported by `sx` prop.
+
+#### Not doing, yet
+
+The following shouldn't be considered right now:
+
+- `className`, `style` props support _(see [drawbacks and constraints](#drawbacksconstraints))_
+- Grid systems
+- Bleed layout components
+- Stack components (same can be achieved with `Flex`, `Grid`)
+- Separate utility component for hiding (same can be achieved with responsive properties on `display` set to `none`, `flex`, `block`, etc.)
+- Page level layout components (nav, header, sidebar, footer, etc.)
+
+### Roadmap
+
+Since layout components have a huge scope we should target to launch these progressively. This also gives us room to iterate based on feedback from consumers.
+
+Here's a proposed timeline:
+
+1. `Box`, `Flex`, `Grid` component
+1. `Spacer` component
+1. Responsive styles
```

Comment:
Shouldn't responsive styles be supported from day 1 for all layout components? Why release this iteratively? Even though this is a frequently asked component, its a non-blocker since folks are already using workarounds. We should instead build this out properly in the first go to avoid API changes in iterations.

I think we should do
1. `Flex` & `Grid`(web only) components with responsive styles support
2. `Spacer` component with responsive styles support
3. `sx` prop with responsive styles support

`useSx`, we can take it up later if valid use-cases come in

Reply:
This can definitely work, we can club responsive properties in first iteration.

Reply:
I'm still wondering why do we need these many components? Majorly we need
- Flex/Box(anyone) with responsive capabilities
- spacing props for presentational components eg: Button, Text, etc.

Even for Flex I would propose we can have a general purpose name, something like View as reshaped/spectrume uses it. and then use it for Flex things as well has other props from box like position props, background color, etc.

Reply:
Updated roadmap in https://github.com/razorpay/blade/pull/950/commits/f4e3b6cebb79e8d1a3b72ed54eb53ed67c55dec4

Current proposal removes these helper components from our side. Though it has slight side effect on readability, listed in drawbacks.